### ISSUE TITLE
configure some iterations to run in the same job?

### ISSUE THREAD

**Issue by tdhock** [mlr-org/mlr3batchmark#32]
Hi @sebffischer 
I was wondering if it is currently possible to run different benchmark iterations in the same cluster job?

In particular I would like to tell mlr3batchmark to create a new job for every data set and cross-validation fold, but have all the different algorithms run one after another in the same job. is that possible?

**Comment by sebffischer** [comment]
If I understand you correctly, this can be achieved by chunking jobs: https://mllg.github.io/batchtools/reference/submitJobs.html#chunking-of-jobs

**Comment by tdhock** [comment]
probably best for me would be if I could specify this as an argument to batchmark 
```r
(bench.grid <- mlr3::benchmark_grid( 
  tasks=task.list,
  learners=learner.list,
  resamplings=train.test.cv))
mlr3batchmark::batchmark(
      bench.grid, store_models = TRUE, reg=reg,
      job.for.each=c("resampling","task")) #could also specify "learner" here
```
does that seem reasonable to you? or would you suggest another approach? If you don't have time to write this functionality, I could give it a try, but it would be useful to have some guidance about what you think the interface should look like /where would be the best place to edit the current code.

**Comment by be-marc** [comment]
Why do you want to run multiple resampling iterations in one job?

**Comment by tdhock** [comment]
some clusters have limits on the number of jobs/tasks that can be running/queued simultaneously. 
For example most clusters in compute canada have a limit of 1000.
For example if I run featureless baseline learner, and one other algorithm, I can combine the two into a single task so that the total number of tasks is less than 1000. 
Does that make sense to you?

**Comment by be-marc** [comment]
Yes, I understand the problem. In this case, you would normally set `max.concurrent.jobs = 1000` in `batchtools.conf.R`.
`submitJobs()` would then wait after submitting 1000 jobs to the queue. When 1 jobs is finished, it sends the job 1001  to the queue. If you can`t stay active on the login node, I would write a small script that  executes `submitJobs()`  in a job.

**Comment by be-marc** [comment]
or another option: Why don't you merge more jobs into one array job? As far as I know, they are then counted as one job on a Slurm cluster. If you use `chunks.as.arrayjobs=TRUE` then you should have already aggregated the jobs in some form already.

**Comment by tdhock** [comment]
I did not know about max.concurrent.jobs, that is helpful.
I'm not sure about staying active on login node, so that solution is less than ideal.

> Why don't you merge more jobs into one array job? As far as I know, they are then counted as one job on a Slurm cluster. 

If I understand your suggestion correctly, you think I could assign all 1200 batchtools jobs in a single slurm job with 1200 tasks? Actually on our cluster I think each task in the job array is counted toward the limit of 1000, so that solution would not work.

I believe that is implemented via https://slurm.schedmd.com/resource_limits.html#assoc_maxsubmitjobs but that page does not specify how job arrays are handled.

**Comment by be-marc** [comment]
Okay then use `max.concurrent.jobs = 1000` and start `submitJobs()` in a long running job itself.

**Comment by tdhock** [comment]
thanks for the quick feedback! I guess that could be a work-around in the short term.
but since that solution is less convenient for my use case, I would still like to fork mlr3batchmark and try to hack my proposal. could you please share any guidance about how/where to implement my proposition?
or would adding that feature not be an acceptable addition to mlr3batchmark?

**Comment by be-marc** [comment]
I don't think this requires a change to mlr3batchtmark but to [batchtools](https://github.com/mllg/batchtools/). @mllg (maintainer) Do you think this could be useful?

**Comment by berndbischl** [comment]
ok, this is clearly BT issue, but let me answer here, after our productive discussion with TH in M.

1) the current system is SLIGHTLY illdefined. chunking and arayjobs should be 2 independent concepts, but the current system mixes them, slightly hackily.

2) chunking should always be sequential chunking. array jobs should just be a way of referring to BT jobs. 

3) in the template file our R code "doJobCollection" is called.  the path to the RDS file normally determines in the job collection it o execute. in AJ the path in the doJobColletion should be NULL. the path CANNOT determine the which jobs are executed, this is determined by the envir var of AJ. 

4) I think we have the id numbers of AJ undercontrol so we simply set the required values there, and in doJobCollection we load the envir var, and from THAT we call the correct jobcollection

**Comment by tdhock** [comment]
The proposed user interface that we discussed is as follows.

The jobs table already can have `chunk` column, we propose adding `array` column.

The interpretation is: for each unique value of array, create a SLURM job by running sbatch with --array=1-(number of unique chunk values).

Consider a job table with 4 rows (each row is one batchtools job).

Use case 1: neither chunk nor array specified. Defaults are chunk=array=c(1,2,3,4). 4 SLURM jobs, each runs 1 task = 1 batchtools job.

Use case 2: array=c(1,1,2,2) specified by user, chunk defaults to c(1,2,3,4). 2 SLURM jobs, 2 tasks each, each task runs 1 batchtools job.

Use case 3: chunk=c(1,1,2,2) specified by user, array defaults to c(1,1,2,2). 2 SLURM jobs (1 task each), each runs 2 batchtools jobs in sequence.

Use case 4: array=c(1,1,1,1) and chunk=c(1,1,2,2) specified by user. 1 SLURM job, 2 tasks, each runs 2 batchtools jobs in sequence.

Use case 5: array=c(1,1,1,1) specified by user, chunk defaults to c(1,2,3,4). 1 SLURM job, 4 tasks, each runs 1 batchtools job.

General case:
- if chunk not specified, it defaults to 1:N
- if array not specified, it defaults to chunk.

**Comment by berndbischl** [comment]
design suggestion

registry: contains all BT jobs 1...m

chunking this is optional
it will create jobcollections.
if chunking is not used JCs are also
there (then of size 1)
the file layout on disk is now
jc_1.rds, ...., jc_k.rds
NB: k can be different than m
if chunking is enabled.
point is: jc_i.rds simply contains the 
job BT ids, which are suppposed to be executed sequentially.

submitJobs happens. 
this takes a DT [id | chunk].
this clear defines the chunks, and which
BT jobs the chunk contains. 
SJ now ACTUALLY defined the jc_i.rds files
and files them to disk.

now look at the slurm template file.

it contains the line
doJobCollection(path to JC file)

case A:
no array jobs. we simply execute the JC file, so we run all contained BT jobs, 
with their contained IDS, sequentially.

case B: 
array jobs. the path to the JC is NULL.
in doJobCollection, we read the envir var.
then execute that JC file, as in case A.

we need the array job number in the table
so we have DT [job | JC | array]

so SJ now:
creates the job script
it looks like this
case A: 
doJobCollection(path)

case B: 
#SBATCH --array=<task ids>
doJobCollection(NULL)

now, task IDS are simply all of the 
JC ids, which refer to the same array.


----------
remarks: 
i1) talking about chunks and JCs
seems nonsensical. it is ALWAYS JCs.
use a single name.
i2) usally, users dont want to FREELY
control id nrs of arrays, the simply want
arrays of a certain size. but this is now
just talking about "sugar"

**Comment by tdhock** [comment]
So the root problem of this issue was the fact that my SLURM clusters in Canada have a max of 1000 jobs running or in the queue at a time. This is problematic for large benchmark experiments, say 20 data sets * 10-fold CV * 10 learning algos = 2000 jobs, which is over the limit of 1000.

Original post asked for ability to combine algorithms into the same job, which in the example amounts to 200 jobs, each which runs one data set, one split, 10 algos (ok because under limit of 1000). Would be nice if batchmark could support this use case.

I coded up an alternative solution in https://github.com/tdhock/mlr3resampling/pull/32

* `mlr3resampling::proj_grid()` is a replacement for `mlr3::benchmark_grid()` that writes a project directory with some files that describe the different benchmark iterations.
* `mlr3resampling::proj_submit()` uses batchtools to sbatch a SLURM job array with a certain number of tasks (max 1000)
* each one of those tasks calls `mlr3resampling::proj_compute_until_done()` which looks at a file to see what benchmark iterations are left to do, and starts computing one, then when done, looks again, etc. So each task does a variable number of benchmark iterations. The different iterations in each task are essentially random (not all the algos for a given task/split as I original asked), but that is OK.
* finally when there is no more work to do, `mlr3resampling::proj_results_save()` is called to combine all individual results into a single RDS result file which contains all results.

I did an experiment on the cluster https://github.com/tdhock/mlr3resampling/pull/32#issuecomment-2929426787 with 5 learners, 10-fold CV, 2 test subsets, same/other/all train subsets (300 benchmark iterations total).

* `batchmark()` makes one SLURM job array with 300 tasks, each with time limit of 24 hours.
* `mlr3resampling::proj_grid()` etc: I asked for one SLURM job array with 100 tasks, each with time limit of 24 hours.
* I observed that I can get the same results using either method, so 3x more efficient use of cluster resources via `mlr3resampling::proj_grid()` etc.

I would be grateful if you could review that PR, if you have time/interest.
